pkg/registry/core/service/storage/storage_test.go

```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
+		}
+	}
+}
+
+type cudTestCase struct {
+	name   string
+	create svcTestCase
+	update svcTestCase
+}
+
+func helpTestCreateUpdateDelete(t *testing.T, testCases []cudTestCase) {
+	// NOTE: do not call t.Helper() here.  It's more useful for errors to be
+	// attributed to lines in this function than the caller of it.
+
+	// This test is ONLY with the gate enabled.
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.IPv6DualStack, true)()
+
+	storage, _, server := newStorage(t, []api.IPFamily{api.IPv4Protocol, api.IPv6Protocol})
+	defer server.Terminate(t)
+	defer storage.Store.DestroyFunc()
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := genericapirequest.NewDefaultContext()
+
+			// Create the object as specified and check the results.
+			obj, err := storage.Create(ctx, tc.create.svc, rest.ValidateAllObjectFunc, &metav1.CreateOptions{})
+			if tc.create.expectError && err != nil {
+				return
+			}
+			if err != nil {
+				t.Fatalf("unexpected error creating service: %v", err)
+			}
+			defer storage.Delete(ctx, tc.create.svc.Name, rest.ValidateAllObjectFunc, &metav1.DeleteOptions{}) // in case
+			if tc.create.expectError && err == nil {
+				t.Fatalf("unexpected success creating service")
+			}
+			createdSvc := obj.(*api.Service)
+			if !verifyEquiv(t, "create", &tc.create, createdSvc) {
+				return
+			}
+			if tc.create.expectClusterIPs {
+				proveClusterIPsAllocated(t, storage, tc.create.svc, createdSvc)
+			} else if tc.create.expectHeadless {
+				proveHeadless(t, storage, tc.create.svc, createdSvc)
+			} else {
+				proveClusterIPsDeallocated(t, storage, tc.create.svc, createdSvc)
+			}
+			if tc.create.expectNodePorts {
+				proveNodePortsAllocated(t, storage, tc.create.svc, createdSvc)
+			} else {
+				proveNodePortsDeallocated(t, storage, tc.create.svc, createdSvc)
+			}
+			if tc.create.expectHealthCheckNodePort {
+				proveHealthCheckNodePortAllocated(t, storage, tc.create.svc, createdSvc)
+			} else {
+				proveHealthCheckNodePortDeallocated(t, storage, tc.create.svc, createdSvc)
+			}
+
+			// Update the object to the new state and check the results.
+			obj, created, err := storage.Update(ctx, tc.update.svc.Name,
+				rest.DefaultUpdatedObjectInfo(tc.update.svc), rest.ValidateAllObjectFunc,
+				rest.ValidateAllObjectUpdateFunc, false, &metav1.UpdateOptions{})
+			if tc.update.expectError && err != nil {
+				return
+			}
+			if err != nil {
+				t.Fatalf("unexpected error updating service: %v", err)
+			}
+			if tc.update.expectError && err == nil {
+				t.Fatalf("unexpected success updating service")
+			}
+			if created {
+				t.Fatalf("unexpected create-on-update")
+			}
+			updatedSvc := obj.(*api.Service)
+			if !verifyEquiv(t, "update", &tc.update, updatedSvc) {
+				return
+			}
+			if tc.update.expectClusterIPs {
```

Comment:
Commit `Service REST: feature tests`: 
It is basically the same call as L6051 define an inlined func and let us call it twice.

Reply:
ACK.